src/kraken.rs

```diff
@@ -0,0 +1,139 @@
+use crate::kraken::btc_dai::{MidMarketRate, Rate};
+use serde::de::Error;
+use serde::Deserialize;
+use std::convert::TryFrom;
+
+/// Fetch Ticker data
+/// More info here: https://www.kraken.com/features/api
+pub async fn get_mid_market_rate() -> anyhow::Result<MidMarketRate> {
+    let rate = reqwest::get("https://api.kraken.com/0/public/Ticker?pair=XBTDAI")
+        .await?
+        .json::<TickerResponse>()
+        .await
+        .map(|response| response.result.xbtdai)?;
+
+    Ok(MidMarketRate::from_kraken(rate))
+}
+
+#[derive(Deserialize)]
+struct TickerResponse {
+    result: Ticker,
+}
+
+#[derive(Deserialize)]
+struct Ticker {
+    #[serde(rename = "XBTDAI")]
+    xbtdai: Rate,
+}
+
+#[derive(Deserialize)]
+struct TickerData {
+    #[serde(rename = "a")]
+    ask: Vec<String>,
+    #[serde(rename = "b")]
+    bid: Vec<String>,
+}
+
+impl TryFrom<TickerData> for Rate {
+    type Error = serde_json::Error;
+
+    fn try_from(value: TickerData) -> Result<Self, Self::Error> {
+        let ask_price = value
+            .ask
+            .first()
```

Comment:
Is `first` (returned by Kraken) the most recent or the highest? We are looking for the highest price, aren't we?

Reply:
AFAIK, the ask field of a trading ticker contains the latest information. Same for the bid.

Here `.first()` is just selecting the first field in the array `[price, whole_lot_volume, lot_volume]`, according to https://www.kraken.com/features/api.